activerecord/lib/active_record/relation/query_methods.rb

```diff
@@ -53,7 +53,7 @@ def not(opts, *rest)
     Relation::MULTI_VALUE_METHODS.each do |name|
       class_eval <<-CODE, __FILE__, __LINE__ + 1
         def #{name}_values                    # def select_values
-          @values[:#{name}] || []             #   @values[:select] || []
+          @values[:#{name}] ||= []            #   @values[:select] ||= []
```

Comment:
Here's what I went with:
```ruby
require 'cases/helper'
require 'models/post'

require 'benchmark/ips'

module ActiveRecord
  class RecordFetchWarningTest < ActiveRecord::TestCase
    fixtures :posts

    def test_bench
      permutations = (0..3).flat_map { |i| ([true] * i + [false] * (3 - i)).permutation.to_a.uniq }

      Benchmark.ips do |bm|
        permutations.each do |permutation|
          $set_array, $set_clause, $set_hash = *permutation

          relation = Post.all
          relation.load
          second_relation = Post.all
          second_relation.load

          bm.report("spawn: set_array: #{$set_array}, set_clause: #{$set_clause}, set_hash: #{$set_hash}") { relation.spawn }
        end

        bm.compare!
      end

      Benchmark.ips do |bm|
        permutations.each do |permutation|
          $set_array, $set_clause, $set_hash = *permutation

          relation = Post.all
          relation.load
          second_relation = Post.all
          second_relation.load

          bm.report("merge: set_array: #{$set_array}, set_clause: #{$set_clause}, set_hash: #{$set_hash}") { relation.merge(second_relation) }
        end

        bm.compare!
      end
    end
  end
end
```
And I applied this diff:
```diff
diff --git a/activerecord/lib/active_record/relation/query_methods.rb b/activerecord/lib/active_record/relation/query_methods.rb
index 867bc16..0b830a5 100644
--- a/activerecord/lib/active_record/relation/query_methods.rb
+++ b/activerecord/lib/active_record/relation/query_methods.rb
@@ -53,7 +53,12 @@ module ActiveRecord
     Relation::MULTI_VALUE_METHODS.each do |name|
       class_eval <<-CODE, __FILE__, __LINE__ + 1
         def #{name}_values                    # def select_values
-          @values[:#{name}] ||= []            #   @values[:select] ||= []
+          # @values[:#{name}] ||= []            #   @values[:select] ||= []
+          if $set_array
+            @values[:#{name}] ||= []
+          else
+            @values[:#{name}] || []
+          end
         end                                   # end
                                               #
         def #{name}_values=(values)           # def select_values=(values)
@@ -83,7 +88,12 @@ module ActiveRecord
     Relation::CLAUSE_METHODS.each do |name|
       class_eval <<-CODE, __FILE__, __LINE__ + 1
         def #{name}_clause                           # def where_clause
-          @values[:#{name}] ||= new_#{name}_clause   #   @values[:where] ||= new_where_clause
+          # @values[:#{name}] ||= new_#{name}_clause   #   @values[:where] ||= new_where_clause
+          if $set_clause
+            @values[:#{name}] ||= new_#{name}_clause
+          else
+            @values[:#{name}] || new_#{name}_clause
+          end
         end                                          # end
                                                      #
         def #{name}_clause=(value)                   # def where_clause=(value)
@@ -98,7 +108,12 @@ module ActiveRecord
     end
 
     def create_with_value # :nodoc:
-      @values[:create_with] ||= {}
+      # @values[:create_with] ||= {}
+      if $set_hash
+        @values[:create_with] ||= {}
+      else
+        @values[:create_with] || {}
+      end
     end
 
     alias extensions extending_values

```
As you say, there is a slowdown when copying:
```
Calculating -------------------------------------
spawn: set_array: false, set_clause: false, set_hash: false
                        28.367k i/100ms
spawn: set_array: true, set_clause: false, set_hash: false
                        16.709k i/100ms
spawn: set_array: false, set_clause: true, set_hash: false
                        23.045k i/100ms
spawn: set_array: false, set_clause: false, set_hash: true
                        27.183k i/100ms
spawn: set_array: true, set_clause: true, set_hash: false
                        15.163k i/100ms
spawn: set_array: true, set_clause: false, set_hash: true
                        16.910k i/100ms
spawn: set_array: false, set_clause: true, set_hash: true
                        20.742k i/100ms
spawn: set_array: true, set_clause: true, set_hash: true
                        14.802k i/100ms
-------------------------------------------------
spawn: set_array: false, set_clause: false, set_hash: false
                        534.965k (± 7.8%) i/s -      2.666M
spawn: set_array: true, set_clause: false, set_hash: false
                        221.189k (±13.4%) i/s -      1.103M
spawn: set_array: false, set_clause: true, set_hash: false
                        364.490k (± 7.8%) i/s -      1.821M
spawn: set_array: false, set_clause: false, set_hash: true
                        452.526k (± 9.0%) i/s -      2.256M
spawn: set_array: true, set_clause: true, set_hash: false
                        189.651k (±12.2%) i/s -    940.106k
spawn: set_array: true, set_clause: false, set_hash: true
                        212.384k (±14.8%) i/s -      1.032M
spawn: set_array: false, set_clause: true, set_hash: true
                        356.481k (±12.0%) i/s -      1.763M
spawn: set_array: true, set_clause: true, set_hash: true
                        195.456k (±11.6%) i/s -    962.130k

Comparison:
spawn: set_array: false, set_clause: false, set_hash: false:   534964.8 i/s
spawn: set_array: false, set_clause: false, set_hash: true:   452526.5 i/s - 1.18x slower
spawn: set_array: false, set_clause: true, set_hash: false:   364489.5 i/s - 1.47x slower
spawn: set_array: false, set_clause: true, set_hash: true:   356481.4 i/s - 1.50x slower
spawn: set_array: true, set_clause: false, set_hash: false:   221188.5 i/s - 2.42x slower
spawn: set_array: true, set_clause: false, set_hash: true:   212383.9 i/s - 2.52x slower
spawn: set_array: true, set_clause: true, set_hash: true:   195456.0 i/s - 2.74x slower
spawn: set_array: true, set_clause: true, set_hash: false:   189650.9 i/s - 2.82x slower

Calculating -------------------------------------
merge: set_array: false, set_clause: false, set_hash: false
                         1.306k i/100ms
merge: set_array: true, set_clause: false, set_hash: false
                         1.277k i/100ms
merge: set_array: false, set_clause: true, set_hash: false
                         1.571k i/100ms
merge: set_array: false, set_clause: false, set_hash: true
                         1.061k i/100ms
merge: set_array: true, set_clause: true, set_hash: false
                         1.512k i/100ms
merge: set_array: true, set_clause: false, set_hash: true
                         1.354k i/100ms
merge: set_array: false, set_clause: true, set_hash: true
                         1.740k i/100ms
merge: set_array: true, set_clause: true, set_hash: true
                         1.512k i/100ms
-------------------------------------------------
merge: set_array: false, set_clause: false, set_hash: false
                         14.672k (± 7.1%) i/s -     73.136k
merge: set_array: true, set_clause: false, set_hash: false
                         13.485k (±10.2%) i/s -     67.681k
merge: set_array: false, set_clause: true, set_hash: false
                         18.798k (± 6.8%) i/s -     94.260k
merge: set_array: false, set_clause: false, set_hash: true
                         15.079k (± 7.2%) i/s -     75.331k
merge: set_array: true, set_clause: true, set_hash: false
                         16.701k (±13.8%) i/s -     81.648k
merge: set_array: true, set_clause: false, set_hash: true
                         12.918k (±14.9%) i/s -     63.638k
merge: set_array: false, set_clause: true, set_hash: true
                         17.527k (± 9.2%) i/s -     87.000k
merge: set_array: true, set_clause: true, set_hash: true
                         15.934k (± 9.6%) i/s -     80.136k

Comparison:
merge: set_array: false, set_clause: true, set_hash: false:    18797.8 i/s
merge: set_array: false, set_clause: true, set_hash: true:    17527.2 i/s - 1.07x slower
merge: set_array: true, set_clause: true, set_hash: false:    16701.1 i/s - 1.13x slower
merge: set_array: true, set_clause: true, set_hash: true:    15934.1 i/s - 1.18x slower
merge: set_array: false, set_clause: false, set_hash: true:    15079.0 i/s - 1.25x slower
merge: set_array: false, set_clause: false, set_hash: false:    14672.3 i/s - 1.28x slower
merge: set_array: true, set_clause: false, set_hash: false:    13485.2 i/s - 1.39x slower
merge: set_array: true, set_clause: false, set_hash: true:    12917.5 i/s - 1.46x slower
```

How shall we proceed?